### PR TITLE
Added throttle for message publishing

### PR DESCRIPTION
**Pull request by pierantoniomerlino**
The idea is to manage the rate and burst of publishing using the Token Bucket algorithm.
The user can configure in the DataService the "Maximum publish rate" and "Maximum number of messages in publish bursts". The first parameter is used to set the maximum messages per seconds that can be published. The second one, instead, sets the number of messages that can be published in a burst at max speed. After the burst, the publishing rate is limited to "Maximum publish rate" value.

If the "Maximum publish rate" is set to 0, no limits are set and the messaged are published at max speed. If "Maximum number of messages in publish bursts" is set to 0, the burst are disabled and the messages are published at "Maximum publish rate". The latter parameter is used only if the "Maximum publish rate" is set to a value different than 0.

Of course, this PR is preliminary and any suggestions are more than welcome!

Signed-off-by: pierantoniomerlino <pierantonio.merlino@eurotech.com>

### CODE REVIEW

File: kura/org.eclipse.kura.core/OSGI-INF/metatype/org.eclipse.kura.data.DataService.xml [hunk 1]
```diff
@@ -98,7 +98,25 @@
             default="0"
             min="0"
             description="Timeouts the in-flight messages congestion condition. The service will force a disconnect attempting to reconnect (0 to disable)."/>
+
+        <AD id="max.publish.rate"
+            name="Maximum publish rate"
+            type="Double"
```
Comment (MMaiero) [inline kura/org.eclipse.kura.core/OSGI-INF/metatype/org.eclipse.kura.data.DataService.xml]: Why this is a double?

File: kura/org.eclipse.kura.core/src/main/java/org/eclipse/kura/core/data/DataServiceImpl.java [hunk 2]
```diff
@@ -104,6 +109,14 @@ protected void activate(ComponentContext componentContext, Map<String, Object> p
 
         this.properties.putAll(properties);
 
+        if (this.properties.get(PUBLISH_RATE) != null && this.properties.get(BURST_LENGTH) != null) {
```
Comment (MMaiero) [inline kura/org.eclipse.kura.core/src/main/java/org/eclipse/kura/core/data/DataServiceImpl.java]: I think that this can be externalized to a separate method.

File: kura/org.eclipse.kura.core/OSGI-INF/metatype/org.eclipse.kura.data.DataService.xml (hunk 1 above)
Comment (pierantoniomerlino) [reply kura/org.eclipse.kura.core/OSGI-INF/metatype/org.eclipse.kura.data.DataService.xml]: max.publish.rate is the number of messages that should be send in a second. So, if you want to publish 1 messages every 10 seconds, the value is 0.1. Float is sufficient.
Is there a better way to show it?

Comment (MMaiero) [reply kura/org.eclipse.kura.core/OSGI-INF/metatype/org.eclipse.kura.data.DataService.xml]: Maybe specifying a time unit.

File: kura/org.eclipse.kura.core/src/main/java/org/eclipse/kura/core/data/DataServiceImpl.java [hunk 3]
```diff
@@ -526,6 +536,27 @@ public void run() {
         return autoConnect;
     }
 
+    private void getThrottle() {
+        if (this.properties.get(PUBLISH_RATE) != null && this.properties.get(BURST_LENGTH) != null) {
```
Comment (MMaiero) [inline kura/org.eclipse.kura.core/src/main/java/org/eclipse/kura/core/data/DataServiceImpl.java]: Isn't it more readable if we save `this.properties.get(PUBLISH_RATE)` and `this.properties.get(BURST_LENGTH)` into local variables?

**Comment by MMaiero** [comment]
If we compare our implementation to a token bucket, I think that maybe we should stick with common terms like (https://en.wikipedia.org/wiki/Token_bucket): 

- `Average rate` instead of `Maximum publish rate`

- `Burst size` instead of `Maximum number of messages in publish bursts`

What do you think?

File: kura/org.eclipse.kura.core/src/main/java/org/eclipse/kura/core/data/DataServiceImpl.java (hunk 3 above)
Comment (pierantoniomerlino) [reply kura/org.eclipse.kura.core/src/main/java/org/eclipse/kura/core/data/DataServiceImpl.java]: 👍

**Comment by pierantoniomerlino** [comment]
IMHO, we have to think what is easier to understand from an user point of view.
Is it better to expose the "token bucket" terms or use terms related to the data service?

File: kura/org.eclipse.kura.core/src/main/java/org/eclipse/kura/core/data/DataServiceImpl.java [hunk 4]
```diff
@@ -572,14 +598,7 @@ public void run() {
...
 
-                        publishInternal(message);
-
-                        // TODO: add a 'message throttle' configuration parameter to
-                        // slow down publish rate?
-
-                        // Notify the listeners
-                        DataServiceImpl.this.dataServiceListeners.onMessagePublished(message.getId(),
-                                message.getTopic());
+                        tryPublishInternal(message);
```
Comment (cdealti) [inline kura/org.eclipse.kura.core/src/main/java/org/eclipse/kura/core/data/DataServiceImpl.java]: why "try"?

File: kura/org.eclipse.kura.core/src/main/java/org/eclipse/kura/core/data/DataServiceImpl.java [hunk 5]
```diff
@@ -593,6 +612,15 @@ public void run() {
         });
     }
 
+    private synchronized void tryPublishInternal(DataMessage message) throws KuraException {
+        if (this.throttle.getToken()) {
+            logger.debug("Throttle: got token - remaining token {}", this.throttle.getRemainingTokens());
```
Comment (cdealti) [inline kura/org.eclipse.kura.core/src/main/java/org/eclipse/kura/core/data/DataServiceImpl.java]: "remaining tokens"

File: kura/org.eclipse.kura.core/src/main/java/org/eclipse/kura/core/data/internal/TokenBucket.java [hunk 6]
```diff
@@ -0,0 +1,79 @@
+/*******************************************************************************
+ * Copyright (c) 2011, 2017 Eurotech and/or its affiliates
```
Comment (cdealti) [inline kura/org.eclipse.kura.core/src/main/java/org/eclipse/kura/core/data/internal/TokenBucket.java]: only 2017

File: kura/org.eclipse.kura.core/src/main/java/org/eclipse/kura/core/data/internal/TokenBucket.java [hunk 7]
```diff
@@ -0,0 +1,79 @@
...
+ *
+ * All rights reserved. This program and the accompanying materials
+ * are made available under the terms of the Eclipse Public License v1.0
+ * which accompanies this distribution, and is available at
+ * http://www.eclipse.org/legal/epl-v10.html
+ *
+ * Contributors:
+ *     Eurotech
+ *******************************************************************************/
+package org.eclipse.kura.core.data.internal;
```
Comment (cdealti) [inline kura/org.eclipse.kura.core/src/main/java/org/eclipse/kura/core/data/internal/TokenBucket.java]: I believe it should be org.eclipse.kura.core.internal.data. Check the Wires implementation bundles and the Eclipse guidelines for package naming conventions.

File: kura/org.eclipse.kura.core/src/main/java/org/eclipse/kura/core/data/internal/TokenBucket.java [hunk 8]
```diff
@@ -0,0 +1,79 @@
...
+    private long remainingTokens;
+    // period in mills between 1 token refill
+    private long refillPeriod;
+    private long lastRefillTime;
+
+    public TokenBucket(long capacity, long refillPeriod) {
+        this.capacity = (capacity == 0) ? 1 : capacity;
+        this.remainingTokens = capacity;
+        this.refillPeriod = refillPeriod;
+        this.lastRefillTime = 0L;
```
Comment (cdealti) [inline kura/org.eclipse.kura.core/src/main/java/org/eclipse/kura/core/data/internal/TokenBucket.java]: Remove "L"

File: kura/org.eclipse.kura.core/src/main/java/org/eclipse/kura/core/data/internal/TokenBucket.java [hunk 9]
```diff
@@ -0,0 +1,79 @@
...
+        this.refillPeriod = refillPeriod;
+    }
+
+    public long getLastRefill() {
+        return this.lastRefillTime;
+    }
+
+    public boolean getToken() {
+        boolean success;
+        if (this.refillPeriod == 0L) {
```
Comment (cdealti) [inline kura/org.eclipse.kura.core/src/main/java/org/eclipse/kura/core/data/internal/TokenBucket.java]: Remove "L"

File: kura/org.eclipse.kura.core/OSGI-INF/metatype/org.eclipse.kura.data.DataService.xml [hunk 10]
```diff
@@ -98,7 +98,36 @@
...
+            min="0"
+            description="The average message publish rate. The unit of measurement is set by the publish rate unit of measurement option. If set to 0 the rate limit is disabled."/>
+        
+        <AD id="average.publish.rate.unit"
+            name="publish rate unit of measurement"
+            type="String"
+            cardinality="0" 
+            required="true"
+            default="messages.second"
+            description="">
```
Comment (cdealti) [inline kura/org.eclipse.kura.core/OSGI-INF/metatype/org.eclipse.kura.data.DataService.xml]: Add description

File: kura/org.eclipse.kura.core/src/main/java/org/eclipse/kura/core/data/internal/TokenBucket.java [hunk 11]
```diff
@@ -0,0 +1,79 @@
...
+ * http://www.eclipse.org/legal/epl-v10.html
+ *
+ * Contributors:
+ *     Eurotech
+ *******************************************************************************/
+package org.eclipse.kura.core.data.internal;
+
+public class TokenBucket {
+
+    private long capacity;
```
Comment (cdealti) [inline kura/org.eclipse.kura.core/src/main/java/org/eclipse/kura/core/data/internal/TokenBucket.java]: There will plenty of room just using ``int``.

File: kura/org.eclipse.kura.core/src/main/java/org/eclipse/kura/core/data/internal/TokenBucket.java [hunk 12]
```diff
@@ -0,0 +1,79 @@
...
+    private long remainingTokens;
```
Comment (cdealti) [inline kura/org.eclipse.kura.core/src/main/java/org/eclipse/kura/core/data/internal/TokenBucket.java]: Same

File: kura/org.eclipse.kura.core/src/main/java/org/eclipse/kura/core/data/internal/TokenBucket.java [hunk 13]
```diff
@@ -0,0 +1,79 @@
...
+    // period in mills between 1 token refill
+    private long refillPeriod;
+    private long lastRefillTime;
+
+    public TokenBucket(long capacity, long refillPeriod) {
+        this.capacity = (capacity == 0) ? 1 : capacity;
```
Comment (cdealti) [inline kura/org.eclipse.kura.core/src/main/java/org/eclipse/kura/core/data/internal/TokenBucket.java]: Unclear

File: kura/org.eclipse.kura.core/src/main/java/org/eclipse/kura/core/data/internal/TokenBucket.java [hunk 14]
```diff
@@ -0,0 +1,79 @@
...
+        this.remainingTokens = capacity;
+        this.refillPeriod = refillPeriod;
+        this.lastRefillTime = 0L;
+    }
+
+    public long getCapacity() {
+        return this.capacity;
+    }
+
+    public void setCapacity(long capacity) {
```
Comment (cdealti) [inline kura/org.eclipse.kura.core/src/main/java/org/eclipse/kura/core/data/internal/TokenBucket.java]: Remove setters. Create a new instance instead

File: kura/org.eclipse.kura.core/src/main/java/org/eclipse/kura/core/data/internal/TokenBucket.java [hunk 15]
```diff
@@ -0,0 +1,79 @@
...
+
+    public void setRefillPeriod(long refillPeriod) {
+        this.refillPeriod = refillPeriod;
+    }
+
+    public long getLastRefill() {
+        return this.lastRefillTime;
+    }
+
+    public boolean getToken() {
```
Comment (cdealti) [inline kura/org.eclipse.kura.core/src/main/java/org/eclipse/kura/core/data/internal/TokenBucket.java]: ``hasToken`` preferred

File: kura/org.eclipse.kura.core/src/main/java/org/eclipse/kura/core/data/internal/TokenBucket.java [hunk 16]
```diff
@@ -0,0 +1,79 @@
...
+        boolean success;
```
Comment (cdealti) [inline kura/org.eclipse.kura.core/src/main/java/org/eclipse/kura/core/data/internal/TokenBucket.java]: intialize to ``false``

File: kura/org.eclipse.kura.core/src/main/java/org/eclipse/kura/core/data/internal/TokenBucket.java [hunk 17]
```diff
@@ -0,0 +1,79 @@
...
+        if (this.refillPeriod == 0L) {
+            success = true;
+        } else {
+            tryRefill();
+            if (this.remainingTokens > 0) {
+                this.remainingTokens--;
+                success = true;
+            } else {
```
Comment (cdealti) [inline kura/org.eclipse.kura.core/src/main/java/org/eclipse/kura/core/data/internal/TokenBucket.java]: remove ``else``

File: kura/org.eclipse.kura.core/src/main/java/org/eclipse/kura/core/data/internal/TokenBucket.java [hunk 18]
```diff
@@ -0,0 +1,79 @@
...
+                success = false;
+            }
+        }
+        return success;
+    }
+
+    private void tryRefill() {
```
Comment (cdealti) [inline kura/org.eclipse.kura.core/src/main/java/org/eclipse/kura/core/data/internal/TokenBucket.java]: Just ``refill`` returning the number of remaining tokens.

However, we miss an ``waitToken`` method which blocks until one token is available.
Make sure that the wait can be properly interrupted.

File: kura/org.eclipse.kura.core/src/main/java/org/eclipse/kura/core/data/DataServiceImpl.java [hunk 19]
```diff
@@ -156,6 +164,8 @@ public void updated(Map<String, Object> properties) {
         this.properties.clear();
         this.properties.putAll(properties);
 
+        getThrottle();
```
Comment (cdealti) [inline kura/org.eclipse.kura.core/src/main/java/org/eclipse/kura/core/data/DataServiceImpl.java]: ``createThrottle`` or something indicating that an instanciation  will occur

File: kura/org.eclipse.kura.core/src/main/java/org/eclipse/kura/core/data/DataServiceImpl.java [hunk 20]
```diff
@@ -526,6 +536,28 @@ public void run() {
         return autoConnect;
     }
 
+    private void getThrottle() {
+        if (this.properties.get(PUBLISH_RATE) != null && this.properties.get(BURST_SIZE) != null) {
```
Comment (cdealti) [inline kura/org.eclipse.kura.core/src/main/java/org/eclipse/kura/core/data/DataServiceImpl.java]: Hence a new ``TokenBucket`` instance will be created

File: kura/org.eclipse.kura.core/OSGI-INF/metatype/org.eclipse.kura.data.DataService.xml [hunk 21]
```diff
@@ -98,7 +98,36 @@
...
+        <AD id="average.publish.rate"
+            name="Average publish rate"
+            type="Integer"
+            cardinality="0"
+            required="true"
+            default="0"
+            min="0"
+            description="The average message publish rate. The unit of measurement is set by the publish rate unit of measurement option. If set to 0 the rate limit is disabled."/>
+        
+        <AD id="average.publish.rate.unit"
```
Comment (cdealti) [inline kura/org.eclipse.kura.core/OSGI-INF/metatype/org.eclipse.kura.data.DataService.xml]: I believe this is not needed.
I believe that messages/s is the right unit.
Realistically 1 [message/s] is probably what one will set.
Is it too fast for the broker? Setting 0.1 will do the trick

File: kura/org.eclipse.kura.core/OSGI-INF/metatype/org.eclipse.kura.data.DataService.xml [hunk 22]
```diff
@@ -98,7 +98,36 @@
...
+        </AD>
+        
+        <AD id="publish.burst.size"
+            name="Message burst size"
+            type="Long"
+            cardinality="0"
+            required="true"
+            default="0"
+            min="0"
+            description="When the average publish rate is set, this option enables the publishing of messages in bursts at maximum speed. If set to 0 the bursts are disabled."/>
```
Comment (cdealti) [inline kura/org.eclipse.kura.core/OSGI-INF/metatype/org.eclipse.kura.data.DataService.xml]: Just to double check. The broker will not be happy with bursts. Setting this property to 1 disables bursts entirely?

File: kura/org.eclipse.kura.core/src/main/java/org/eclipse/kura/core/data/DataServiceImpl.java [hunk 23]
```diff
@@ -593,6 +618,15 @@ public void run() {
         });
     }
 
+    private synchronized void tryPublishInternal(DataMessage message) throws KuraException {
+        if (this.throttle.getToken()) {
```
Comment (cdealti) [inline kura/org.eclipse.kura.core/src/main/java/org/eclipse/kura/core/data/DataServiceImpl.java]: It should just be ``this.throttle.waitToken()``. Otherwise if this method is called from the publishing task and the DB is full and you don't have tokens, the CPU will spin doing nothing.

With waitToken() at least the thread will block until a token is available.

**Comment by cdealti** [comment]
About the priority of this one. It seems needed to limit the publish rate with Azure IoT but we need to recheck. See #1228.

@markoer  the DataService is a key component that if it fails will cause headaches with customers.
Assuming this will be merged, can you please add some new QA tests for validation covering the corner cases? You should pay attention to cases where the messages DB contains many messages, like after a disconnect and the client reconnects.

File: kura/org.eclipse.kura.core/src/main/java/org/eclipse/kura/core/data/DataServiceImpl.java (hunk 4 above)
Comment (pierantoniomerlino) [reply kura/org.eclipse.kura.core/src/main/java/org/eclipse/kura/core/data/DataServiceImpl.java]: Because if there aren't tokens in the bucket, the Data service can't publish the message. So, it tries to publish something...

File: kura/org.eclipse.kura.core/src/main/java/org/eclipse/kura/core/data/DataServiceImpl.java (hunk 5 above)
Comment (pierantoniomerlino) [reply kura/org.eclipse.kura.core/src/main/java/org/eclipse/kura/core/data/DataServiceImpl.java]: 👍

File: kura/org.eclipse.kura.core/src/main/java/org/eclipse/kura/core/data/internal/TokenBucket.java (hunk 6 above)
Comment (pierantoniomerlino) [reply kura/org.eclipse.kura.core/src/main/java/org/eclipse/kura/core/data/internal/TokenBucket.java]: 👍

File: kura/org.eclipse.kura.core/src/main/java/org/eclipse/kura/core/data/internal/TokenBucket.java (hunk 13 above)
Comment (pierantoniomerlino) [reply kura/org.eclipse.kura.core/src/main/java/org/eclipse/kura/core/data/internal/TokenBucket.java]: I'll add a comment about this.

File: kura/org.eclipse.kura.core/OSGI-INF/metatype/org.eclipse.kura.data.DataService.xml (hunk 21 above)
Comment (pierantoniomerlino) [reply kura/org.eclipse.kura.core/OSGI-INF/metatype/org.eclipse.kura.data.DataService.xml]: Yes, but is it enough clear to the user set a rate of 0.1 message/second? If the maximum rate is 1 message/second, so we should change the unit to message/minute and get an integer value.

File: kura/org.eclipse.kura.core/OSGI-INF/metatype/org.eclipse.kura.data.DataService.xml (hunk 22 above)
Comment (pierantoniomerlino) [reply kura/org.eclipse.kura.core/OSGI-INF/metatype/org.eclipse.kura.data.DataService.xml]: The burst size is the capacity of the token bucket. If it is 0, the token bucket constructor force the capacity to 1, so only a token per period can be extracted from the bucket and there will not be bursts.

File: kura/org.eclipse.kura.core/src/main/java/org/eclipse/kura/core/data/DataServiceImpl.java (hunk 23 above)
Comment (pierantoniomerlino) [reply kura/org.eclipse.kura.core/src/main/java/org/eclipse/kura/core/data/DataServiceImpl.java]: Do we need a timeout for this?

File: kura/org.eclipse.kura.core/OSGI-INF/metatype/org.eclipse.kura.data.DataService.xml (hunk 21 above)
Comment (cdealti) [reply kura/org.eclipse.kura.core/OSGI-INF/metatype/org.eclipse.kura.data.DataService.xml]: I suggest to add the 0.1 example to the description

File: kura/org.eclipse.kura.core/OSGI-INF/metatype/org.eclipse.kura.data.DataService.xml (hunk 22 above)
Comment (cdealti) [reply kura/org.eclipse.kura.core/OSGI-INF/metatype/org.eclipse.kura.data.DataService.xml]: Then set min="1". WYSIWYG

File: kura/org.eclipse.kura.core/OSGI-INF/metatype/org.eclipse.kura.data.DataService.xml (hunk 21 above)
Comment (pierantoniomerlino) [reply kura/org.eclipse.kura.core/OSGI-INF/metatype/org.eclipse.kura.data.DataService.xml]: I'll remove the unit option and set the rate as messages/second. I'll also update the description to give an example on how to use it. However, I don't like it.

File: kura/org.eclipse.kura.core/OSGI-INF/metatype/org.eclipse.kura.data.DataService.xml (hunk 22 above)
Comment (pierantoniomerlino) [reply kura/org.eclipse.kura.core/OSGI-INF/metatype/org.eclipse.kura.data.DataService.xml]: Ok.

**Comment by cdealti** [comment]
See #1416.

File: kura/org.eclipse.kura.core/src/main/java/org/eclipse/kura/core/data/DataServiceImpl.java [hunk 24]
```diff
@@ -76,7 +70,7 @@
     private ScheduledFuture<?> reconnectFuture;
 
     // A dedicated executor for the publishing task
-    private ScheduledExecutorService publisherExecutor;
+    private ExecutorService publisherExecutor;
```
Comment (cdealti) [inline kura/org.eclipse.kura.core/src/main/java/org/eclipse/kura/core/data/DataServiceImpl.java:74]: It looks to me that a plain old Java Thread would me enough

File: kura/org.eclipse.kura.core/OSGI-INF/metatype/org.eclipse.kura.data.DataService.xml [hunk 25]
```diff
@@ -98,7 +98,46 @@
             default="0"
             min="0"
             description="Timeouts the in-flight messages congestion condition. The service will force a disconnect attempting to reconnect (0 to disable)."/>
-            
+
+		<AD id="enable.rate.limit"
```
Comment (cdealti) [inline kura/org.eclipse.kura.core/OSGI-INF/metatype/org.eclipse.kura.data.DataService.xml]: Fix indentation

File: kura/org.eclipse.kura.core/src/main/java/org/eclipse/kura/core/data/DataServiceImpl.java [hunk 26]
```diff
@@ -516,16 +523,27 @@ public void run() {
...
-                    TimeUnit.SECONDS);
+            }, initialDelay, reconnectInterval, TimeUnit.SECONDS);
         } else {
             // Change notification status to off. Connection is not expected to happen in the future
             this.cloudConnectionStatusService.updateStatus(this, CloudConnectionStatusEnum.OFF);
         }
         return autoConnect;
     }
 
+    private void createThrottle() {
```
Comment (cdealti) [inline kura/org.eclipse.kura.core/src/main/java/org/eclipse/kura/core/data/DataServiceImpl.java:533]: Should we add NPE checks for this.throttle everywhere?

File: kura/org.eclipse.kura.core/src/main/java/org/eclipse/kura/core/data/DataServiceImpl.java [hunk 27]
```diff
@@ -680,4 +650,99 @@ public CloudConnectionStatusEnum getNotificationStatus() {
...
+                        DataMessage message = DataServiceImpl.this.store.getNextMessage();
+
+                        if (message != null) {
+                            checkInFlightMessages(message);
+
+                            boolean messagePublished;
+                            if (dataServiceOptions.isRateLimitEnabled() && message.getPriority() >= 5) {
+                                messagePublished = publishMessageTokenBucket(message);
+                            } else {
+                                messagePublished = publishMessageUnbound(message);
```
Comment (cdealti) [inline kura/org.eclipse.kura.core/src/main/java/org/eclipse/kura/core/data/DataServiceImpl.java]: this method alway returns true so set messagePublished to true explicitly or intialize it to true.

File: kura/org.eclipse.kura.core/src/main/java/org/eclipse/kura/core/data/DataServiceImpl.java [hunk 28]
```diff
@@ -680,4 +650,99 @@ public CloudConnectionStatusEnum getNotificationStatus() {
...
+                    && DataServiceImpl.this.inFlightMsgIds.size() >= dataServiceOptions.getMaxInFlightMessages()) {
+                logger.warn("The configured maximum number of in-flight messages has been reached");
+                throw new KuraTooManyInflightMessagesException("Too many in-flight messages");
+            }
+        }
+
+        private void suspendPublisher(long timeout, TimeUnit timeUnit) throws InterruptedException {
+            try {
+                lock.lock();
+                if (!publisherEnabled) {
```
Comment (cdealti) [inline kura/org.eclipse.kura.core/src/main/java/org/eclipse/kura/core/data/DataServiceImpl.java]: Why ``publisherEnabled`` must be checked in a lock?
You don't need to throw an Exception, just return.

File: kura/org.eclipse.kura.core/src/main/java/org/eclipse/kura/core/data/DataServiceImpl.java [hunk 29]
```diff
@@ -680,4 +650,99 @@ public CloudConnectionStatusEnum getNotificationStatus() {
...
+                        lockCondition.await(timeout, timeUnit);
+                    }
+                }
+                notifyPending = false;
+            } finally {
+                lock.unlock();
+            }
+        }
+
+        private boolean publishMessageUnbound(DataMessage message) throws KuraException {
```
Comment (cdealti) [inline kura/org.eclipse.kura.core/src/main/java/org/eclipse/kura/core/data/DataServiceImpl.java]: return void

File: kura/org.eclipse.kura.core/src/main/java/org/eclipse/kura/core/data/DataServiceOptions.java [hunk 30]
```diff
@@ -0,0 +1,117 @@
+package org.eclipse.kura.core.data;
```
Comment (cdealti) [inline kura/org.eclipse.kura.core/src/main/java/org/eclipse/kura/core/data/DataServiceOptions.java:12]: Copyright header missing

File: kura/org.eclipse.kura.core/src/main/java/org/eclipse/kura/core/internal/data/TokenBucket.java [hunk 31]
```diff
@@ -0,0 +1,58 @@
...
+ *
+ * Contributors:
+ *     Eurotech
+ *******************************************************************************/
+package org.eclipse.kura.core.internal.data;
+
+public class TokenBucket {
+
+    private final int capacity;
+    // period in mills between 1 token refill
```
Comment (cdealti) [inline kura/org.eclipse.kura.core/src/main/java/org/eclipse/kura/core/internal/data/TokenBucket.java]: Check spelling or remove comment

File: kura/org.eclipse.kura.core/src/main/java/org/eclipse/kura/core/internal/data/TokenBucket.java [hunk 32]
```diff
@@ -0,0 +1,58 @@
...
+        boolean result = false;
+        refill();
+        if (isTokenAvailable()) {
+            this.remainingTokens--;
+            result = true;
+        }
+        return result;
+    }
+
+    private boolean isTokenAvailable() {
```
Comment (cdealti) [inline kura/org.eclipse.kura.core/src/main/java/org/eclipse/kura/core/internal/data/TokenBucket.java:38]: Move private methods at the bottom

File: kura/org.eclipse.kura.core/src/main/java/org/eclipse/kura/core/internal/data/TokenBucket.java [hunk 33]
```diff
@@ -0,0 +1,58 @@
...
+        return this.remainingTokens != 0;
+    }
+
+    private void refill() {
+        long now = System.currentTimeMillis();
+        if (now - this.lastRefillTime >= this.refillPeriod) {
+            this.remainingTokens = (int) Math.min(this.capacity,
+                    this.remainingTokens + (now - this.lastRefillTime) / this.refillPeriod);
+            this.remainingTokens = Math.max(1, this.remainingTokens);
+            this.lastRefillTime = now;
```
Comment (cdealti) [inline kura/org.eclipse.kura.core/src/main/java/org/eclipse/kura/core/internal/data/TokenBucket.java]: lastRefillTime is not accurate as you lose the elapsed part of the refillPeriod at the time refill() is called.
I believe it should be computed as:

lastRefillTime += ((now - lastRefillTime) / refillPeriod) * refillPeriod

Note that this does not equal ``now`` due to the calculation uses integer arithmetics.

File: kura/org.eclipse.kura.core/src/main/java/org/eclipse/kura/core/internal/data/TokenBucket.java [hunk 34]
```diff
@@ -0,0 +1,58 @@
...
+        }
+    }
+
+    public long getNextRefillEta() {
```
Comment (cdealti) [inline kura/org.eclipse.kura.core/src/main/java/org/eclipse/kura/core/internal/data/TokenBucket.java]: Maybe add a comment that you don't get an updated and meaningful value until getToken() is called. This usage is misleading and error prone however.
The method name should also indicate a "time to next refill" (relative to now) instead of an "ETA of the next refill" (relative to epoch).

**Review by cdealti — CHANGES_REQUESTED** [review]

File: kura/org.eclipse.kura.core/src/main/java/org/eclipse/kura/core/data/DataServiceImpl.java (hunk 24 above)
Comment (MMaiero) [reply kura/org.eclipse.kura.core/src/main/java/org/eclipse/kura/core/data/DataServiceImpl.java:74]: I believe in this context there is no real advantage in one or the other solution.

File: kura/org.eclipse.kura.core/src/main/java/org/eclipse/kura/core/data/DataServiceImpl.java (hunk 28 above)
Comment (MMaiero) [reply kura/org.eclipse.kura.core/src/main/java/org/eclipse/kura/core/data/DataServiceImpl.java]: It's to prevent concurrent changes and reads of the shared variable.

File: kura/org.eclipse.kura.core/src/main/java/org/eclipse/kura/core/data/DataServiceOptions.java (hunk 30 above)
Comment (MMaiero) [reply kura/org.eclipse.kura.core/src/main/java/org/eclipse/kura/core/data/DataServiceOptions.java:12]: Correct. Missed that.

File: kura/org.eclipse.kura.core/src/main/java/org/eclipse/kura/core/data/DataServiceImpl.java (hunk 26 above)
Comment (MMaiero) [reply kura/org.eclipse.kura.core/src/main/java/org/eclipse/kura/core/data/DataServiceImpl.java:533]: Uhm, where do you think it is needed?

File: kura/org.eclipse.kura.core/src/main/java/org/eclipse/kura/core/data/DataServiceImpl.java [hunk 35]
```diff
@@ -680,4 +649,106 @@ public CloudConnectionStatusEnum getNotificationStatus() {
...
+                }
+            }
+            logger.debug("Exited publisher loop.");
+        }
+
+        private boolean isExecutionEnabled() {
+            boolean executionEnabled;
+            lock.lock();
+            executionEnabled = publisherEnabled;
+            lock.unlock();
```
Comment (pierantoniomerlino) [inline kura/org.eclipse.kura.core/src/main/java/org/eclipse/kura/core/data/DataServiceImpl.java]: Is it possible to use an AtomicBoolean instead?

Comment (MMaiero) [reply kura/org.eclipse.kura.core/src/main/java/org/eclipse/kura/core/data/DataServiceImpl.java]: Agreed and done. I've done also some minimal testing and it seems to work.
